// ngx-color/circle/esm5/circle.component.js
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { amber, blue, blueGrey, brown, cyan, deepOrange, deepPurple, green, indigo, lightBlue, lightGreen, lime, orange, pink, purple, red, teal, yellow, } from 'material-colors';
import { ColorWrap, SwatchModule } from 'ngx-color';
import { isValidHex } from 'ngx-color';
import { CircleSwatchComponent } from './circle-swatch.component';
var CircleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CircleComponent, _super);
    function CircleComponent() {
        var _this = _super.call(this) || this;
        /** Pixel value for picker width */
        _this.width = 252;
        /** Color squares to display */
        _this.colors = [
            red['500'],
            pink['500'],
            purple['500'],
            deepPurple['500'],
            indigo['500'],
            blue['500'],
            lightBlue['500'],
            cyan['500'],
            teal['500'],
            green['500'],
            lightGreen['500'],
            lime['500'],
            yellow['500'],
            amber['500'],
            orange['500'],
            deepOrange['500'],
            brown['500'],
            blueGrey['500'],
        ];
        /** Value for circle size */
        _this.circleSize = 28;
        /** Value for spacing between circles */
        _this.circleSpacing = 14;
        return _this;
    }
    CircleComponent.prototype.isActive = function (color) {
        return this.hex === color;
    };
    CircleComponent.prototype.handleBlockChange = function (_a) {
        var hex = _a.hex, $event = _a.$event;
        if (isValidHex(hex)) {
            this.handleChange({ hex: hex, source: 'hex' }, $event);
        }
    };
    CircleComponent.prototype.handleValueChange = function (_a) {
        var data = _a.data, $event = _a.$event;
        this.handleChange(data, $event);
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], CircleComponent.prototype, "width", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Array)
    ], CircleComponent.prototype, "colors", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], CircleComponent.prototype, "circleSize", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], CircleComponent.prototype, "circleSpacing", void 0);
    CircleComponent = tslib_1.__decorate([
        Component({
            selector: 'color-circle',
            template: "\n    <div\n      class=\"circle-picker {{ className }}\"\n      [style.width.px]=\"width\"\n      [style.margin-right.px]=\"-circleSpacing\"\n      [style.margin-bottom.px]=\"-circleSpacing\"\n    >\n      <color-circle-swatch\n        *ngFor=\"let color of colors\"\n        [circleSize]=\"circleSize\"\n        [circleSpacing]=\"circleSpacing\"\n        [color]=\"color\"\n        [focus]=\"isActive(color)\"\n        (onClick)=\"handleBlockChange($event)\"\n        (onSwatchHover)=\"onSwatchHover.emit($event)\"\n      ></color-circle-swatch>\n    </div>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush,
            preserveWhitespaces: false,
            styles: ["\n      .circle-picker {\n        display: flex;\n        flex-wrap: wrap;\n      }\n    "]
        }),
        tslib_1.__metadata("design:paramtypes", [])
    ], CircleComponent);
    return CircleComponent;
}(ColorWrap));
export { CircleComponent };
var ColorCircleModule = /** @class */ (function () {
    function ColorCircleModule() {
    }
    ColorCircleModule = tslib_1.__decorate([
        NgModule({
            declarations: [CircleComponent, CircleSwatchComponent],
            exports: [CircleComponent, CircleSwatchComponent],
            imports: [CommonModule, SwatchModule],
        })
    ], ColorCircleModule);
    return ColorCircleModule;
}());
export { ColorCircleModule };
